### PR TITLE
Add new features

### PR DESCRIPTION
**Pull request by MrMeison**
*  Add an ability to freeze the timeline and task when we scroll down or scroll across
* Add an ability to change Gantt progress bar colors

### CODE REVIEW

File: src/gantt.ts [hunk 1]
```diff
@@ -141,29 +136,24 @@ module powerbi.extensibility.visual {
...
-        types: string[];
         typeName: string;
+        types: TaskTypeMetadata[];
     };
 
+    export interface TaskTypeMetadata {
+        name: string;
+        columnGroup: DataViewValueColumnGroup;
+        selectionColumn: DataViewCategoryColumn;
+
```
Comment (ignatvilesov) [inline src/gantt.ts]: Please remove the empty line.

File: src/gantt.ts [hunk 2]
```diff
@@ -365,12 +350,43 @@ module powerbi.extensibility.visual {
...
+            // create axis container
+            this.axisGroup = this.ganttSvg
+                .append("g")
+                .classed(Selectors.AxisGroup.class, true);
+            this.axisGroup
+                .append("rect")
+                .attr("width", "100%")
+                .attr("y", "-20")
+                .attr("height", "40px")
+                .attr("fill", "white");
```
Comment (ignatvilesov) [inline src/gantt.ts]: Is it possible to specify these attributes using CSS?

File: src/gantt.ts [hunk 3]
```diff
@@ -365,12 +350,43 @@ module powerbi.extensibility.visual {
...
+                .attr("height", "100%")
+                .attr("fill", "white");
+
             // create legend container
             this.legend = createLegend(element,
                 this.isInteractiveChart,
                 this.interactivityService,
                 true,
                 LegendPosition.Top);
+            this.ganttDiv.on("scroll", function (evt) {
```
Comment (ignatvilesov) [inline src/gantt.ts]: Does it work on mobile devices with touch screen?

File: src/gantt.ts [hunk 4]
```diff
@@ -497,92 +497,107 @@ module powerbi.extensibility.visual {
...
+                values.Task.forEach((categoryValue: PrimitiveValue, index: number) => {
+                    if (group.Duration && group.Duration.values[index] !== null) {
+                        const selectoinBuider: ISelectionIdBuilder = host
+                            .createSelectionIdBuilder()
+                            .withCategory(dataView.categorical.categories[0], index);
+                        let color = Gantt.DefaultValues.TaskColor;
+                        const taskType = _.find(taskTypes.types, (typeMeta: TaskTypeMetadata) => typeMeta.name === group.Duration.source.groupName);
+                        if (taskType) {
+                            selectoinBuider
+                                .withCategory(taskType.selectionColumn, 0);
```
Comment (ignatvilesov) [inline src/gantt.ts]: Does selection work properly?

File: src/gantt.ts [hunk 5]
```diff
@@ -497,92 +497,107 @@ module powerbi.extensibility.visual {
...
+                            color = colorHelper.getColorForMeasure(taskType.columnGroup.objects, taskType.name);
+                        }
+                        const selectionId = selectoinBuider.createSelectionId();
```
Comment (ignatvilesov) [inline src/gantt.ts]: Please specify a type of the variable.

File: src/gantt.ts [hunk 6]
```diff
@@ -497,92 +497,107 @@ module powerbi.extensibility.visual {
...
+                        const task: Task = {
+                            color,
+                            completion,
+                            resource,
+                            id: index,
+                            name: categoryValue as string,
+                            start: startDate,
+                            end: null,
+                            duration: duration,
+                            taskType: taskType.name as string || "",
```
Comment (ignatvilesov) [inline src/gantt.ts]: Might ```taskType``` be empty or undefined?

File: src/gantt.ts [hunk 7]
```diff
@@ -593,45 +608,22 @@ module powerbi.extensibility.visual {
...
-            });
-
-            const formatters: GanttChartFormatters = this.getFormatters(dataView);
+            const settings: GanttSettings = GanttSettings.parse<GanttSettings>(dataView);
 
-            const tasks: Task[] = Gantt.createTasks(dataView, host, formatters, colors);
-            const series: GanttSeries[] = Gantt.createSeries(dataView, host, tasks, colors);
+            const taskTypes: TaskTypes = Gantt.getAllTasksTypes(dataView)
+                , formatters: GanttChartFormatters = this.getFormatters(dataView)
+                , tasks = Gantt.createTasks(dataView, taskTypes, host, formatters, colors);
```
Comment (ignatvilesov) [inline src/gantt.ts]: Please specify a type of the variable.

File: tslint.json [hunk 8]
```diff
@@ -10,11 +23,13 @@
...
         ],
         "no-duplicate-variable": true,
-        "no-eval": true,
         "no-internal-module": false,
         "no-trailing-whitespace": true,
         "no-unsafe-finally": true,
         "no-var-keyword": true,
+        "no-unused-variable": true,
+        "no-unused-new": true,
+        "no-unused-expression": true,
```
Comment (ignatvilesov) [inline tslint.json]: Please include [rules](https://github.com/Microsoft/PowerBI-visuals-ChicletSlicer/blob/master/tslint.json#L57) that have been recommended by Avi.

File: src/gantt.ts (hunk 3 above)
Comment (MrMeison) [inline src/gantt.ts]: yes, why not? 
https://developer.mozilla.org/en-US/docs/Web/Events/scroll

File: src/gantt.ts (hunk 4 above)
Comment (MrMeison) [inline src/gantt.ts]: Yes, I checked it

File: src/gantt.ts (hunk 6 above)
Comment (MrMeison) [inline src/gantt.ts]: Yes, when the visual hasn't the legend

File: tslint.json (hunk 8 above)
Comment (MrMeison) [inline tslint.json]: Added

File: src/gantt.ts (hunk 2 above)
Comment (MrMeison) [inline src/gantt.ts]: No, `width` and `height` property work no correctly on Edge, if it use as CSS rules

File: package.json [hunk 9]
```diff
@@ -28,34 +27,40 @@
...
-    "powerbi-visuals-utils-svgutils": "^0.2.1",
-    "powerbi-visuals-utils-tooltiputils": "^0.3.0",
-    "powerbi-visuals-utils-typeutils": "^0.2.0",
-    "powerbi-visuals-utils-chartutils": "^0.2.1",
-    "powerbi-visuals-utils-testutils": "^0.2.1",
-    "tslint": "4.0.2",
-    "typings": "2.0.0"
+    "karma-typescript-preprocessor": "0.3.1",
+    "powerbi-visuals-tools": "1.6.3",
+    "powerbi-visuals-utils-testutils": "1.0.0",
```
Comment (AviSander) [inline package.json]: shouldn't this be in "dependencies"?

Comment (MrMeison) [inline package.json]: No, this is `devDependencies`. Those libraries will not be included in result visual package. They use only in development environments such as testing or build process

**Pull request merged**